Mathlib/Algebra/GroupPower/Lemmas.lean

```diff
@@ -2,35 +2,1316 @@
+  rfl
+#align multiples_hom_symm_apply multiples_hom_symm_apply
+
+attribute [to_additive multiples_hom_symm_apply] powers_hom_symm_apply
+
+@[simp]
+theorem zmultiples_hom_apply [AddGroup A] (x : A) (n : ℤ) : zmultiplesHom A x n = n • x :=
+  rfl
+#align zmultiples_hom_apply zmultiples_hom_apply
+
+attribute [to_additive zmultiples_hom_apply] zpowers_hom_apply
+
+@[simp]
+theorem zmultiples_hom_symm_apply [AddGroup A] (f : ℤ →+ A) : (zmultiplesHom A).symm f = f 1 :=
+  rfl
+#align zmultiples_hom_symm_apply zmultiples_hom_symm_apply
+
+attribute [to_additive zmultiples_hom_symm_apply] zpowers_hom_symm_apply
+
+-- TODO use to_additive in the rest of this file
+theorem MonoidHom.apply_mnat [Monoid M] (f : Multiplicative ℕ →* M) (n : Multiplicative ℕ) :
+    f n = f (Multiplicative.ofAdd 1) ^ (Multiplicative.toAdd n) := by
+  rw [← powers_hom_symm_apply, ← powers_hom_apply, Equiv.apply_symm_apply]
+#align monoid_hom.apply_mnat MonoidHom.apply_mnat
+
+@[ext]
+theorem MonoidHom.ext_mnat [Monoid M] ⦃f g : Multiplicative ℕ →* M⦄
+    (h : f (Multiplicative.ofAdd 1) = g (Multiplicative.ofAdd 1)) : f = g :=
+  MonoidHom.ext fun n => by rw [f.apply_mnat, g.apply_mnat, h]
+#align monoid_hom.ext_mnat MonoidHom.ext_mnat
+
+theorem MonoidHom.apply_mint [Group M] (f : Multiplicative ℤ →* M) (n : Multiplicative ℤ) :
+    f n = f (Multiplicative.ofAdd 1) ^ (Multiplicative.toAdd n) := by
+  rw [← zpowers_hom_symm_apply, ← zpowers_hom_apply, Equiv.apply_symm_apply]
+#align monoid_hom.apply_mint MonoidHom.apply_mint
+
+/-! `monoid_hom.ext_mint` is defined in `data.int.cast` -/
+
+
+theorem AddMonoidHom.apply_nat [AddMonoid M] (f : ℕ →+ M) (n : ℕ) : f n = n • f 1 := by
+  rw [← multiples_hom_symm_apply, ← multiples_hom_apply, Equiv.apply_symm_apply]
+#align add_monoid_hom.apply_nat AddMonoidHom.apply_nat
+
+/-! `add_monoid_hom.ext_nat` is defined in `data.nat.cast` -/
+
+
+theorem AddMonoidHom.apply_int [AddGroup M] (f : ℤ →+ M) (n : ℤ) : f n = n • f 1 := by
+  rw [← zmultiples_hom_symm_apply, ← zmultiples_hom_apply, Equiv.apply_symm_apply]
+#align add_monoid_hom.apply_int AddMonoidHom.apply_int
+
+/-! `add_monoid_hom.ext_int` is defined in `data.int.cast` -/
+
+
+variable (M G A)
+
+/-- If `M` is commutative, `powers_hom` is a multiplicative equivalence. -/
+def powersMulHom [CommMonoid M] : M ≃* (Multiplicative ℕ →* M) :=
+  { powersHom M with map_mul' := fun a b => MonoidHom.ext (
+    by
+      intro n
+      let n' : ℕ  := Multiplicative.toAdd n
+      show (a*b) ^ n' = a ^ n' * b ^ n'
+      simp [mul_pow]
+    ) }
+#align powers_mul_hom powersMulHom
+
+/-- If `M` is commutative, `zpowers_hom` is a multiplicative equivalence. -/
+def zpowersMulHom [CommGroup G] : G ≃* (Multiplicative ℤ →* G) :=
+  { zpowersHom G with map_mul' := fun a b => MonoidHom.ext (
+    by
+      intro n
+      let n' : ℤ   := Multiplicative.toAdd n
+      show (a*b) ^ n' = a ^ n' * b ^ n'
+      simp [mul_zpow]
+    )
+  -- <| by simp [mul_zpow]
```

Comment:
add `porting note` and note the regression

Reply:
Done